src/OpenTelemetry.Instrumentation.Dependencies/Implementation/SqlEventSourceListener.netfx.cs

```diff
@@ -0,0 +1,131 @@
+// limitations under the License.
+// </copyright>
+#if NETFRAMEWORK
+using System;
+using System.Data.SqlClient;
+using System.Diagnostics;
+using System.Diagnostics.Tracing;
+using System.Reflection;
+using OpenTelemetry.Trace;
+
+namespace OpenTelemetry.Instrumentation.Dependencies.Implementation
+{
+    /// <summary>
+    /// .NET Framework SqlClient doesn't emit DiagnosticSource events.
+    /// We hook into its EventSource if it is available:
+    /// See: <a href="https://github.com/microsoft/referencesource/blob/3b1eaf5203992df69de44c783a3eda37d3d4cd10/System.Data/System/Data/Common/SqlEventSource.cs#L29">reference source</a>.
+    /// </summary>
+    internal class SqlEventSourceListener : EventListener
+    {
+        internal const string ActivitySourceName = "SqlClient";
+        internal const string ActivityName = ActivitySourceName + ".Execute";
+
+        private static readonly Version Version = typeof(SqlEventSourceListener).Assembly.GetName().Version;
+        private static readonly ActivitySource SqlClientActivitySource = new ActivitySource(ActivitySourceName, Version.ToString());
+        private static readonly EventSource SqlEventSource = (EventSource)typeof(SqlConnection).Assembly.GetType("System.Data.SqlEventSource")?.GetField("Log", BindingFlags.Static | BindingFlags.NonPublic)?.GetValue(null);
+
+        private readonly SqlClientInstrumentationOptions options;
+
+        public SqlEventSourceListener(SqlClientInstrumentationOptions options = null)
+        {
+            this.options = options ?? new SqlClientInstrumentationOptions();
+
+            if (SqlEventSource != null)
+            {
+                this.EnableEvents(SqlEventSource, EventLevel.Informational);
+            }
+        }
+
+        public override void Dispose()
+        {
+            if (SqlEventSource != null)
+            {
+                this.DisableEvents(SqlEventSource);
+            }
+
+            base.Dispose();
+        }
+
+        protected override void OnEventWritten(EventWrittenEventArgs eventData)
+        {
+            if (eventData.EventId == 1)
+            {
+                // BeginExecuteEventId
+
+                var activity = SqlClientActivitySource.StartActivity(ActivityName, ActivityKind.Client);
+                if (activity == null)
+                {
+                    return;
+                }
+
+                activity.DisplayName = (string)eventData.Payload[2];
+
+                if (activity.IsAllDataRequested)
+                {
+                    activity.AddTag(SpanAttributeConstants.ComponentKey, "sql");
+
+                    activity.AddTag(SpanAttributeConstants.DatabaseTypeKey, "sql");
+                    activity.AddTag(SpanAttributeConstants.PeerServiceKey, (string)eventData.Payload[1]);
+                    activity.AddTag(SpanAttributeConstants.DatabaseInstanceKey, (string)eventData.Payload[2]);
+
+                    if (string.IsNullOrEmpty((string)eventData.Payload[3]))
```

Comment:
will this throw if Payload has less than 4 count? Lets do a safety check to ensure payload[3] is fine. (some old version did not provide all the payload , maybe irrelevant now, but lets be on defensive side)

Reply:
I added some defensive checks. Now validates for payload being null, payload not being the expected size. Doesn't validate the types of the payload so casts could blow up. But that seemed like taking it a step too far? LMK